ADBDEV-4907-46: Add check for fopen function result

RekGRpth opened the pull request:
Add check for fopen function result

In some files, the debugging functions did not have a check for what the fopen
function returned, so I added such a check.

This patch also adds a warning when fopen fails, because these are dumping
helper functions and it wouldn't be very nice if they crashed on errors like
opening a file.

RekGRpth committed:
Add check for fopen function result

In some files, the debugging functions did not have a check
for what the fopen function returned, so I added such a check.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63334

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034820

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034813

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034814

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034821

bandetto commented on src/backend/cdb/motion/ic_udpifc.c:
```diff
@@ -882,6 +882,9 @@ dumpTransProtoStats()
 	FILE	   *ofile = fopen(tmpbuf, "w+");
 
+	if (ofile == NULL)
+		return;
```
Maybe also use `elog(WARNING)` about `fopen()` failure?

RekGRpth replied:
Yes, this is possible, but this is just a bug fix for the static analyzer for a function that is never called (see the definition of TRANSFER_PROTOCOL_STATS).

bandetto replied:
Are you sure that `TRANSFER_PROTOCOL_STATS` will remain unused?

RekGRpth replied:
I think there is no need to complicate error correction for a static analyzer.

bandetto replied:
These functions are used for debugging, and if file could not be opened, they will now silently fail. You should use `elog()` here with appropriate log level, since these functions may be used in `TeardownUDPIFCInterconnect()` and others, if `TRANSFER_PROTOCOL_STATS` is defined:
https://github.com/arenadata/gpdb/blob/0d43eea955c0df1c3c8a880b37e3c48d4eedb1a4/src/backend/cdb/motion/ic_udpifc.c#L3534-L3536

RekGRpth replied:
added

bandetto replied:
I think `WARNING` is too high for debugging functions. Maybe use `LOG` or `DEBUG1`?
https://github.com/arenadata/gpdb/blob/83ebbf648dadfafc6dd390c99afeddad7754a0e2/src/include/utils/elog.h#L29-L31

RekGRpth replied:
I copied it from https://github.com/arenadata/gpdb/blob/215d8d8f9fec39300e4eb84977ade4fce040ebc1/src/backend/access/bitmap/bitmaputil.c#L1369

RekGRpth committed:
Add check for fopen function result

In some files, the debugging functions did not have a check for what the fopen
function returned, so I added such a check.

This patch also adds warning when fopen fails.

bandetto commented on src/backend/utils/error/debugutils.c:
```diff
@@ -177,6 +177,9 @@ dotnode(void *node, const char *fname)
 	FILE *ofile = fopen(fname, "w+");
 
+	if (ofile == NULL)
+		return;
```
Why is there no error messages?

RekGRpth replied:
added

RekGRpth committed:
add stderr when fopen fails

RekGRpth committed:
fix indent

andr-sokolov commented on src/backend/utils/error/debugutils.c:
```diff
@@ -233,6 +245,12 @@ void dump_mt_bind(MemTupleBinding *mt_bind, const char *fname)
 
+    if (ofile == NULL)
+    {
+        fprintf(stderr, "dump_mt_bind: can't open \"%s\" for output", fname);
```
I suggest to use ```__func__``` to print function name, because someday the function name may be changed and it may cause to difference between the name and the string

RekGRpth replied:
reworked to elog

RekGRpth committed:
Add check for fopen function result

In some files, the debugging functions did not have a check for what the fopen
function returned, so I added such a check.

This patch also adds a warning when fopen fails, because these are dumping
helper functions and it wouldn't be very nice if they crashed on errors like
opening a file.
